pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,382 @@
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving AWS cloud provider")
+		}
+		pvlc.ebsVolumes = awsCloudProvider
+	}
+	return pvlc.ebsVolumes, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findGCEPDLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.gceCloudProvider == nil {
+		gceCloudProvider, ok := pvlc.cloud.(*gce.GCECloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving GCE cloud provider")
+		}
+		pvlc.gceCloudProvider = gceCloudProvider
+	}
+	return pvlc.gceCloudProvider, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) updateVolume(volume *v1.PersistentVolume) error {
+	glog.V(4).Infof("updating PersistentVolume %s", volume.Name)
+	err := clientretry.RetryOnConflict(UpdatePVLabelsBackoff, func() error {
+		curVol, err := pvlc.kubeClient.Core().PersistentVolumes().Get(volume.Name, metav1.GetOptions{IncludeUninitialized: true})
+		if err != nil {
+			return err
+		}
+		curVol.Labels = volume.Labels
+		curVol.Status.Phase = volume.Status.Phase
+		pvlc.removeInitializer(curVol)
+
+		_, err = pvlc.kubeClient.Core().PersistentVolumes().Update(curVol)
+		return err
+	})
+	if err == nil {
+		glog.V(4).Infof("updated PersistentVolume %s", volume.Name)
+	}
+	return err
+}
+
+func (pvlc *PersistentVolumeLabelController) addInitializationConfig() {
+	pvlInitializer := v1alpha1.Initializer{
+		Name: initializerName,
+		Rules: []v1alpha1.Rule{
+			{
+				APIGroups:   []string{"*"},
+				APIVersions: []string{"*"},
+				Resources:   []string{"persistentvolumes"},
+			},
+		},
+	}
+	pvlInitConfig := v1alpha1.InitializerConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: initialzerConfigName,
+		},
+		Initializers: []v1alpha1.Initializer{pvlInitializer},
+	}
+
+	existingInitConfig, err := pvlc.kubeClient.AdmissionregistrationV1alpha1().InitializerConfigurations().Get(initialzerConfigName, metav1.GetOptions{})
+	if err != nil {
+		// InitializerConfig wasn't found
+		_, err = pvlc.kubeClient.AdmissionregistrationV1alpha1().InitializerConfigurations().Create(&pvlInitConfig)
+		if err != nil {
+			glog.Errorf("Failed to set InitializerConfig: %v", err)
+		}
```

Comment:
If this fails, isn't the whole controller broken?

Also, what if an admin wants to limit where this runs?  Or modifies the initializer?

How do you handle adoption (what about PVCs that weren't previously set up with labels)?